### PR TITLE
Improve parser performance

### PR DESCRIPTION
**Pull request by zrzka**
Signed-off-by: Robert Vojta <rvojta@me.com>

### CODE REVIEW

**Comment by zrzka** [comment]
@TimonPost ready for review. The main change is about reading stdin input into 1k buffer instead of byte by byte. Rest is rather cosmetic & small performance gains (nothing big).

This PR should be squashed, lot of uninteresting commits.

I'll sync `anes` tests in another PR to keep it clear what's this all about.

**Comment by TimonPost** [comment]
I will test it on windows, then I will approve this PR, nice improvement!

File: src/event/source/unix.rs [hunk 1]
```diff
@@ -115,26 +127,24 @@ impl EventSource for UnixInternalEventSource {
...
-                                self.tty_buffer.push(self.tty_fd.read_byte()?);
+                                let read_count =
+                                    self.tty_fd.read(&mut self.tty_buffer, TTY_BUFFER_SIZE)?;
 
-                                let input_available = self
-                                    .poll
-                                    .poll(&mut self.events, Some(Duration::from_secs(0)))
-                                    .map(|x| x > 0)?;
+                                if read_count > 0 {
+                                    let additional_input_available = self
```
[the remark this answers is not in the document]
Comment (zrzka) [reply src/event/source/unix.rs]: It's not about events, but if stdin has more input or not.

File: src/event/timeout.rs [hunk 2]
```diff
@@ -19,22 +19,26 @@ impl PollTimeout {
     ///
     /// It always returns `false` if the initial timeout was set to `None`.
     pub fn elapsed(&self) -> bool {
-        self.timeout
-            .map(|timeout| self.start.elapsed() >= timeout)
-            .unwrap_or(false)
+        if let Some(timeout) = self.timeout {
```
[the remark this answers is not in the document]
Comment (zrzka) [reply src/event/timeout.rs]: Negligible performance improvement that I can rollback this probably.

File: src/event/timeout.rs [hunk 3]
```diff
@@ -19,22 +19,26 @@ impl PollTimeout {
...
+            self.start.elapsed() >= timeout
+        } else {
+            false
+        }
     }
 
     /// Returns the timeout leftover (initial timeout duration - elapsed duration).
     pub fn leftover(&self) -> Option<Duration> {
-        self.timeout.map(|timeout| {
+        if let Some(timeout) = self.timeout {
```
[the remark this answers is not in the document]
Comment (zrzka) [reply src/event/timeout.rs]: Negligible performance improvement that I can rollback this probably.

File: examples/perf.rs [hunk 4]
```diff
@@ -0,0 +1,74 @@
+//
+// time socat -s -b4096 -u OPEN:./test.txt,ignoreeof EXEC:'cargo run --example perf',pty,setsid,ctty
```
[the remark this answers is not in the document]
Comment (zrzka) [reply examples/perf.rs]: I removed it and you didn't refresh changes probably. But I see it resolved = "ignoring".

File: src/event/timeout.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (zrzka) [reply src/event/timeout.rs]: Rollback.

File: src/event/timeout.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (zrzka) [reply src/event/timeout.rs]: Rollback.

File: src/event/source/unix.rs [hunk 5]
```diff
@@ -115,26 +127,24 @@ impl EventSource for UnixInternalEventSource {
                     for event in events_count {
```
[the remark this answers is not in the document]
Comment (zrzka) [reply src/event/source/unix.rs]: Refactored it a bit to make it more readable.

**Comment by zrzka** [comment]
Did address all the comments and refactored `try_read` a bit. There was also a bug in the `additional_input_available` handling, because we weren't checking for the `TTY_TOKEN` (it was just `poll()? > 0` = wrong).

**Pull request merged**